cpp/include/cudf/table/experimental/row_operators.cuh

```diff
@@ -420,6 +419,260 @@ class self_comparator {
 };
 
 }  // namespace lexicographic
+
+namespace equality_hashing {
+
+template <typename Nullate>
+class device_row_comparator {
+  friend class self_eq_comparator;
+
+  /**
+   * @brief Construct a function object for performing equality comparison between the rows of two
+   * tables.
+   *
+   * @param has_nulls Indicates if either input table contains columns with nulls.
+   * @param lhs The first table
+   * @param rhs The second table (may be the same table as `lhs`)
+   * @param nulls_are_equal Indicates if two null elements are treated as equivalent
+   */
+  device_row_comparator(Nullate has_nulls,
+                        table_device_view lhs,
+                        table_device_view rhs,
+                        null_equality nulls_are_equal = null_equality::EQUAL)
+    : lhs{lhs}, rhs{rhs}, nulls{has_nulls}, nulls_are_equal{nulls_are_equal}
+  {
+  }
+
+  /**
+   * @brief Performs an equality comparison between two elements in two columns.
+   *
+   * @tparam Nullate A cudf::nullate type describing how to check for nulls.
+   */
+  class element_comparator {
+   public:
+    /**
+     * @brief Construct type-dispatched function object for comparing equality
+     * between two elements.
+     *
+     * @note `lhs` and `rhs` may be the same.
+     *
+     * @param has_nulls Indicates if either input column contains nulls.
+     * @param lhs The column containing the first element
+     * @param rhs The column containing the second element (may be the same as lhs)
+     * @param nulls_are_equal Indicates if two null elements are treated as equivalent
+     */
+    __device__ element_comparator(Nullate has_nulls,
+                                  column_device_view lhs,
+                                  column_device_view rhs,
+                                  null_equality nulls_are_equal = null_equality::EQUAL)
+      : lhs{lhs}, rhs{rhs}, nulls{has_nulls}, nulls_are_equal{nulls_are_equal}
+    {
+    }
+
+    /**
+     * @brief Compares the specified elements for equality.
+     *
+     * @param lhs_element_index The index of the first element
+     * @param rhs_element_index The index of the second element
+     * @return True if lhs and rhs are equal or if both lhs and rhs are null and nulls are
+     * configured to be considered equal (`nulls_are_equal` == `null_equality::EQUAL`)
+     */
+    template <typename Element, CUDF_ENABLE_IF(cudf::is_equality_comparable<Element, Element>())>
+    __device__ bool operator()(size_type const lhs_element_index,
+                               size_type const rhs_element_index) const noexcept
+    {
+      if (nulls) {
+        bool const lhs_is_null{lhs.is_null(lhs_element_index)};
+        bool const rhs_is_null{rhs.is_null(rhs_element_index)};
+        if (lhs_is_null and rhs_is_null) {
+          return nulls_are_equal == null_equality::EQUAL;
+        } else if (lhs_is_null != rhs_is_null) {
+          return false;
+        }
```

Comment:
This is repeated below. Can we factor this into a function?

Reply:
I tried but it gets ugly. The function cannot return a simple bool. It needs to return whether there was a null and if so, whether it was equal. Refactoring it out into
```c++
      if (not null_compare_equal(lhs, rhs, lhs_element_index, rhs_element_index)) { return false; }
      return equality_compare(lhs.element<Element>(lhs_element_index),
                              rhs.element<Element>(rhs_element_index));
```
didn't do the trick because if the check returned true because both elements were null or because both were valid need to be treated differently. In the former, we skip the subsequent element check and in the latter we don't.

Reply:
How about

```
template <typename T>
thrust::optional<T> column_device_view::optional_element(size_type index){
   return is_valid(index) ? {element<T>(index)} : thrust::nullopt;
}
...
template <typename E>
bool nullable_compare(optional<E> const& lhs, optional<E> const& rhs, null_equality n){
   return (lhs and rhs) ? equality_compare(*lhs, *rhs) : (n == null_equality::EQUAL and !lhs and !rhs);
}
...
return nullable_compare(lhs.optional_element<E>(lhs_index), rhs.optional_element<E>(rhs_indx), nulls_are_equal);
```

Reply:
I guess that wouldn't be good, especially below where you want to check the nulls w/o the values.